### PR TITLE
Group rocksdb.sst.read.micros stat by different user read IOActivity + misc

### PR DESCRIPTION
**Pull request by hx235**
**Context/Summary:**
- Similar to https://github.com/facebook/rocksdb/pull/11288 but for user read such as `Get(), MultiGet(), DBIterator::XXX(), Verify(File)Checksum()`. 
   - For this, I refactored some user-facing `MultiGet` calls in `TransactionBase` and various types of `DB` so that it does not call a user-facing `Get()` but `GetImpl()` for passing the `ReadOptions::io_activity` check (see PR conversation)
   - New user read stats breakdown are guarded by `kExceptDetailedTimers` since measurement shows they have 4-5% regression to the upstream/main.

- Misc
   - More refactoring: with https://github.com/facebook/rocksdb/pull/11288, we complete passing `ReadOptions/IOOptions` to FS level. So we can now replace the previously [added](https://github.com/facebook/rocksdb/pull/9424) `rate_limiter_priority` parameter in `RandomAccessFileReader`'s `Read/MultiRead/Prefetch()` with `IOOptions::rate_limiter_priority`
   - Also, `ReadAsync()` call time is measured in `SST_READ_MICRO` now

**Test:**

- CI fake db crash/stress test
- Microbenchmarking

**Build** `make clean && ROCKSDB_NO_FBCODE=1 DEBUG_LEVEL=0 make -jN db_basic_bench`
- google benchmark version: https://github.com/google/benchmark/commit/604f6fd3f4b34a84ec4eb4db81d842fa4db829cd
- db_basic_bench_base: upstream
- db_basic_bench_pr: db_basic_bench_base + this PR
- asyncread_db_basic_bench_base: upstream + [db basic bench patch for IteratorNext](https://github.com/facebook/rocksdb/compare/main...hx235:rocksdb:micro_bench_async_read)
- asyncread_db_basic_bench_pr: asyncread_db_basic_bench_base + this PR

**Test**

Get
```
TEST_TMPDIR=/dev/shm ./db_basic_bench_{null_stat|base|pr} --benchmark_filter=DBGet/comp_style:0/max_data:134217728/per_key_size:256/enable_statistics:1/negative_query:0/enable_filter:0/mmap:1/threads:1 --benchmark_repetitions=1000
```

Result
```
Base:
2038,2059,2061,2065,2065,2072,2075,2077,2082,2088,2088,2090,2090,2090,2095,2095,2097,2097,2098,2099,2100,2101,2102,2102,2102,2103,2104,2104,2104,2104,2105,2106,2106,2106,2106,2106,2107,2107,2107,2108,2108,2108,2108,2108,2109,2109,2110,2110,2110,2111,2111,2111,2112,2112,2112,2112,2113,2113,2114,2114,2115,2116,2116,2118,2118,2118,2118,2119,2119,2119,2119,2119,2120,2120,2120,2121,2121,2121,2121,2121,2122,2123,2123,2123,2123,2123,2124,2124,2125,2125,2126,2126,2126,2126,2127,2127,2127,2127,2128,2128

PR (0.32 % regression, can be seen as none)
2055,2069,2069,2072,2074,2079,2080,2082,2087,2088,2090,2095,2099,2099,2100,2100,2100,2101,2102,2103,2103,2103,2104,2105,2106,2106,2106,2108,2109,2109,2109,2109,2109,2109,2110,2110,2110,2111,2112,2113,2113,2114,2114,2115,2115,2116,2117,2118,2118,2118,2118,2119,2119,2121,2121,2122,2122,2122,2122,2123,2124,2124,2124,2124,2125,2125,2125,2125,2126,2127,2128,2128,2128,2128,2129,2129,2129,2129,2129,2129,2129,2130,2131,2131,2132,2132,2132,2133,2133,2134,2135,2135,2136,2136,2136,2137,2137,2137,2137,2138
```

AsyncRead
```
TEST_TMPDIR=/dev/shm ./asyncread_db_basic_bench_{base|pr} --benchmark_filter=IteratorNext/comp_style:0/max_data:134217728/per_key_size:256/enable_statistics:1/async_io:1/include_detailed_timers:0 --benchmark_repetitions=1000 > syncread_db_basic_bench_{base|pr}.out
```

Result
```
Base:
1956,1956,1968,1977,1979,1986,1988,1988,1988,1990,1991,1991,1993,1993,1993,1993,1994,1996,1997,1997,1997,1998,1999,2001,2001,2002,2004,2007,2007,2008,

PR (2.3% regression, due to measuring `SST_READ_MICRO` that wasn't measured before):
1993,2014,2016,2022,2024,2027,2027,2028,2028,2030,2031,2031,2032,2032,2038,2039,2042,2044,2044,2047,2047,2047,2048,2049,2050,2052,2052,2052,2053,2053,
```

### CODE REVIEW

File: java/rocksjni/portal.h [hunk 1]
```diff
@@ -5627,6 +5627,19 @@ class HistogramTypeJni {
         return 0x3A;
       case ROCKSDB_NAMESPACE::Histograms::FILE_READ_COMPACTION_MICROS:
         return 0x3B;
+      case ROCKSDB_NAMESPACE::Histograms::FILE_READ_DB_OPEN_MICROS:
```
Comment (hx235) [inline java/rocksjni/portal.h:5630]: Rebase conflict with https://github.com/facebook/rocksdb/commit/113f3250f183307793340d0e3b3c075a662648ce

File: utilities/transactions/pessimistic_transaction.cc [hunk 2]
```diff
@@ -166,11 +166,6 @@ template <typename TValue>
 inline Status WriteCommittedTxn::GetForUpdateImpl(
     const ReadOptions& read_options, ColumnFamilyHandle* column_family,
     const Slice& key, TValue* value, bool exclusive, const bool do_validate) {
-  if (read_options.io_activity != Env::IOActivity::kUnknown) {
-    return Status::InvalidArgument(
-        "Cannot call GetForUpdate with `ReadOptions::io_activity` != "
-        "`Env::IOActivity::kUnknown`");
-  }
```
Comment (hx235) [inline utilities/transactions/pessimistic_transaction.cc:173]: Cleanup: remove read options check on TXNs as I intended to add check to db.h level (e.g, TXN db) API only.

File: db/db_impl/db_impl.cc [hunk 3]
```diff
@@ -1927,11 +1927,18 @@ Status DBImpl::Get(const ReadOptions& read_options,
...
+    complete_read_options.io_activity = Env::IOActivity::kGet;
+  }
+
   GetImplOptions get_impl_options;
   get_impl_options.column_family = column_family;
   get_impl_options.value = value;
   get_impl_options.timestamp = timestamp;
-  Status s = GetImpl(read_options, key, get_impl_options);
+
+  Status s = GetImpl(complete_read_options, key, get_impl_options);
```
Comment (hx235) [inline db/db_impl/db_impl.cc:1941]: [Design] The reason why I decided to change `read_options` above `GetImpl()` (or above `VerifyChecksumInternal()` is other db.h level API call (e.g, `  virtual bool KeyMayExist(const ReadOptions& options,
`)can also call `GetImpl()`

File: db/db_impl/db_impl.cc [hunk 4]
```diff
@@ -1948,12 +1955,6 @@ Status DBImpl::GetEntity(const ReadOptions& read_options,
         "Cannot call GetEntity without a PinnableWideColumns object");
   }
 
-  if (read_options.io_activity != Env::IOActivity::kUnknown) {
-    return Status::InvalidArgument(
-        "Cannot call GetEntity with `ReadOptions::io_activity` != "
-        "`Env::IOActivity::kUnknown`");
-  }
-
```
Comment (hx235) [inline db/db_impl/db_impl.cc:1956]: [Note] Please ignore removal of check like this for GetEntity API or Multi/GetForUpdate in TXN related code API. I will either add back the check or include these API as enum in this PR.

**Comment by hx235** [comment]
@ajkr For the purpose of speeding up the review, let me know if you have any high-level design/review concern/comment while this PR is still being stress-tested (the result so far is good)

File: java/rocksjni/portal.h (hunk 1 above)
Comment (hx235) [inline java/rocksjni/portal.h:5630]: Rebase on https://github.com/facebook/rocksdb/pull/11455

File: utilities/transactions/transaction_base.cc [hunk 5]
```diff
@@ -253,8 +248,12 @@ Status TransactionBaseImpl::Get(const ReadOptions& read_options,
...
                                 ColumnFamilyHandle* column_family,
                                 const Slice& key, PinnableSlice* pinnable_val) {
-  return write_batch_.GetFromBatchAndDB(db_, read_options, column_family, key,
-                                        pinnable_val);
+  ReadOptions complete_read_options(read_options);
+  if (complete_read_options.io_activity == Env::IOActivity::kUnknown) {
+    complete_read_options.io_activity = Env::IOActivity::kGet;
+  }
+  return write_batch_.GetFromBatchAndDB(db_, complete_read_options,
+                                        column_family, key, pinnable_val);
```
Comment (ajkr) [inline utilities/transactions/transaction_base.cc:256]: OK I can see why input validation needed to change based on this call. However it looks like it could have changed from:

```
  if (read_options.io_activity != Env::IOActivity::kUnknown) {
// error
```

to

```
  const IOActivity kMyIOActivity = kGet;
  if (read_options.io_activity != Env::IOActivity::kUnknown &&
      read_options.io_activity != kMyIOActivity) {
// error
```

Although this may be irrelevant depending on the answer to the `InternalReadOptions` question

**Review by ajkr — COMMENTED** [review]
We should think about how to do the performance testing. Some regression is expected, but we should know where it happens. For example, does it require statistics enabled and a particular `StatsLevel` to be detectable? In my experience, microbenchmarks  work great for answering these yes/no questions (when run in a targeted way and in a stable environment)

File: utilities/transactions/transaction_base.cc (hunk 5 above)
Comment (hx235) [reply utilities/transactions/transaction_base.cc]: Fixed

File: db_stress_tool/no_batched_ops_stress.cc [hunk 6]
```diff
@@ -578,6 +578,8 @@ class NonBatchedOpsStressTest : public StressTest {
     bool do_consistency_check = thread->rand.OneIn(4);
 
     ReadOptions readoptionscopy = read_opts;
+    readoptionscopy.io_activity = Env::IOActivity::kMultiGet;
```
[the remark this answers is not in the document]
Comment (hx235) [reply db_stress_tool/no_batched_ops_stress.cc]: Fixed

File: db/db_impl/db_impl.cc [hunk 7]
```diff
@@ -1927,11 +1927,18 @@ Status DBImpl::Get(const ReadOptions& read_options,
                    PinnableSlice* value, std::string* timestamp) {
   assert(value != nullptr);
   value->Reset();
+
+  ReadOptions complete_read_options(read_options);
+  if (complete_read_options.io_activity == Env::IOActivity::kUnknown) {
+    complete_read_options.io_activity = Env::IOActivity::kGet;
+  }
```
[the remark this answers is not in the document]
Comment (hx235) [reply db/db_impl/db_impl.cc:1941]: Discussed offline that we might keep io_activity in RO and add option validation in user-facing function instead

File: db/db_impl/compacted_db_impl.cc [hunk 8]
```diff
@@ -46,15 +46,16 @@ Status CompactedDBImpl::Get(const ReadOptions& options, ColumnFamilyHandle*,
 Status CompactedDBImpl::Get(const ReadOptions& options, ColumnFamilyHandle*,
                             const Slice& key, PinnableSlice* value,
                             std::string* timestamp) {
-  if (options.io_activity != Env::IOActivity::kUnknown) {
-    return Status::InvalidArgument(
-        "Cannot call Get with `ReadOptions::io_activity` != "
-        "`Env::IOActivity::kUnknown`");
```
[the remark this answers is not in the document]
Comment (hx235) [reply db/db_impl/compacted_db_impl.cc]: Fixed

File: db/db_impl/db_impl.cc [hunk 9]
```diff
@@ -1999,11 +2006,6 @@ Status DBImpl::GetImpl(const ReadOptions& read_options, const Slice& key,
 
   assert(get_impl_options.column_family);
 
-  if (read_options.io_activity != Env::IOActivity::kUnknown) {
-    return Status::InvalidArgument(
-        "Cannot call Get with `ReadOptions::io_activity` != "
-        "`Env::IOActivity::kUnknown`");
-  }
```
[the remark this answers is not in the document]
Comment (hx235) [reply db/db_impl/db_impl.cc:2006]: Fixed, along with others. I also refactored transaction_base a bit so user-facing function Get() is not called by other user-facing function like GetForUpdate or MultiGet

**Comment by hx235** [comment]
> In my experience, microbenchmarks work great for answering these yes/no questions (when run in a targeted way and in a stable environment)

Updated the test plan with microbench result on Get - as discussed offline, the default stat level will include the new stats reported and A/B on both stats enabled  brings ~3% regression/

**Comment by hx235** [comment]
Fixing the txn UT failures now!

**Comment by ajkr** [comment]
Why do the A/A tests show the lowest p-values (zero)? I thought those tests are running the same binary and the same workload.

**Comment by hx235** [comment]
> Why do the A/A tests show the lowest p-values (zero)? I thought those tests are running the same binary and the same workload.

Might have been a result of me greedily running all 4 comparison the same time - let me try one by one and update the results.

A/A with stats
```
DBGet/comp_style:0/max_data:134217728/per_key_size:256/enable_statistics:1/negative_query:0/enable_filter:0/mmap:1/iterations:1048576/threads:1_pvalue                 0.3722          0.3722      U Test, Repetitions: 80 vs 80
DBGet/comp_style:0/max_data:134217728/per_key_size:256/enable_statistics:1/negative_query:0/enable_filter:0/mmap:1/iterations:1048576/threads:1_mean                  +0.0121         +0.0121          6829          6911          6828          6910
DBGet/comp_style:0/max_data:134217728/per_key_size:256/enable_statistics:1/negative_query:0/enable_filter:0/mmap:1/iterations:1048576/threads:1_median                +0.0135         +0.0135          6693          6784          6692          6782
DBGet/comp_style:0/max_data:134217728/per_key_size:256/enable_statistics:1/negative_query:0/enable_filter:0/mmap:1/iterations:1048576/threads:1_stddev                +0.1554         +0.1556           422           487           421           487
DBGet/comp_style:0/max_data:134217728/per_key_size:256/enable_statistics:1/negative_query:0/enable_filter:0/mmap:1/iterations:1048576/threads:1_cv                    +0.1416         +0.1417             0             0             0             0
OVERALL_GEOMEAN                                                                                                                                                       +0.0115 
```

**Comment by ajkr** [comment]
> Might have been a result of me greedily running all 4 comparison the same time - let me try one by one and update the results.

Then they probably interfere with each other. On further thought I decided to ignore A/A test as it mostly tells us whether there was background noise in that particular run. For A/B test I would take a result as true if it consistently showed low pvalue for the same trend (improvement or regression); however, I don't see any A/B results in your comment.

**Comment by ajkr** [comment]
> however, I don't see any A/B results in your comment

Regarding the ones in the test plan, one result shows an improvement with low pvalue, and the other shows a regression with low pvalue. This seems unlikely to reproduce as it is logically conflicting, but if it does let me know. I'd rather use --benchmark_repetitions=10; if there is a clear regression or improvement it should show regularly.

**Comment by hx235** [comment]
> > Might have been a result of me greedily running all 4 comparison the same time - let me try one by one and update the results.
> 
> Then they probably interfere with each other. On further thought I decided to ignore A/A test as it mostly tells us whether there was background noise in that particular run. For A/B test I would take a result as true if it consistently showed low pvalue for the same trend (improvement or regression) OR if it regularly but not always showed high pvalue (no change); however, I don't see any A/B results in your comment.

Hi - sorry I got distracted from A/B by getting the A/A work so that's why you didn't see any A/B result in this comment

**Comment by hx235** [comment]
> > however, I don't see any A/B results in your comment
> 
> Regarding the ones in the test plan, one result shows an improvement with low pvalue, and the other shows a regression with low pvalue. This seems unlikely to reproduce as it is logically conflicting, but if it does let me know. I'd rather use --benchmark_repetitions=10; if there is a clear regression or improvement it should show regularly.

Sounds good - I will pursue this path then.

**Comment by hx235** [comment]
> > Might have been a result of me greedily running all 4 comparison the same time - let me try one by one and update the results.
> 
> Then they probably interfere with each other. On further thought I decided to ignore A/A test as it mostly tells us whether there was background noise in that particular run. For A/B test I would take a result as true if it consistently showed low pvalue for the same trend (improvement or regression) OR if it regularly but not always showed high pvalue (no change); however, I don't see any A/B results in your comment.

@ajkr 
tldr, 
(1) A/B with stat: if we look at high p-value (blue colored), the mean of performance decreased by 3%
(2) A/B with no stat,  if we look at high p-value (blue colored), the mean of performance decreased by 4 - 5%						

See more in https://docs.google.com/spreadsheets/d/1SwRFpPJhe6216xIUXMRfy5esr6oo9vdPmoIBQKgjHKM/edit?usp=sharing

**Comment by ajkr** [comment]
(1) (stats enabled) I believe ~2% regression since min values I see in your doc are 4652 vs. 4755
(2) (stats disabled) I believe no regression since min values I see in your doc are 3808 vs. 3796

**Comment by hx235** [comment]
@ajkr to make it even more confusing, here is what I got:

Command:
- Build: source /opt/rh/gcc-toolset-11/enable && make -j56 USE_LTO=1 OPTIMIZE_LEVEL="-O3" ROCKSDB_NO_FBCODE=1 DEBUG_LEVEL=0 EXTRA_CXXFLAGS="-I/data/users/huixiao/rocksdb/benchmark/include/" EXTRA_LDFLAGS="-L/data/users/huixiao/rocksdb/benchmark/build/src/" V=1 db_basic_bench
  - db_basic_bench_base is built on https://github.com/facebook/rocksdb/pull/11498
  - db_basic_bench_pr is built on https://github.com/facebook/rocksdb/pull/11498 + https://github.com/facebook/rocksdb/pull/11444
  - db_basic_bench_null_stat (turning off SST_READ_MICROS) is built on https://github.com/facebook/rocksdb/pull/11498 + 
```
/data/users/huixiao/rocksdb$ git diff
diff --git a/file/random_access_file_reader.h b/file/random_access_file_reader.h
index ab4d1e797..7f5dbe652 100644
--- a/file/random_access_file_reader.h
+++ b/file/random_access_file_reader.h
@@ -122,7 +122,7 @@ class RandomAccessFileReader {
       std::unique_ptr<FSRandomAccessFile>&& raf, const std::string& _file_name,
       SystemClock* clock = nullptr,
       const std::shared_ptr<IOTracer>& io_tracer = nullptr,
-      Statistics* stats = nullptr,
+      Statistics* /* stats */ = nullptr,
       uint32_t hist_type = Histograms::HISTOGRAM_ENUM_MAX,
       HistogramImpl* file_read_hist = nullptr,
       RateLimiter* rate_limiter = nullptr,
@@ -132,7 +132,7 @@ class RandomAccessFileReader {
       : file_(std::move(raf), io_tracer, _file_name),
         file_name_(std::move(_file_name)),
         clock_(clock),
-        stats_(stats),
+        stats_(nullptr),
         hist_type_(hist_type),
         file_read_hist_(file_read_hist),
         rate_limiter_(rate_limiter),
```
      - 
- Run (only with stats enabled): TEST_TMPDIR=/dev/shm ./db_basic_bench_{null_stat|base|pr} --benchmark_filter=DBGet/comp_style:0/max_data:134217728/per_key_size:256/enable_statistics:1/negative_query:0/enable_filter:0/mmap:1/threads:1 --benchmark_repetitions=100

Comparison of the top 10 min of each db_basic_bench_*:
db_basic_bench_null_stat:
2452, 2477, 2525, 2609, 2614, 2629, 2670, 2675, 2682, 2684
db_basic_bench_base:
2533, 2651, 2669, 2673, 2678, 2683, 2691, 2697, 2703, 2707
db_basic_bench_pr:
2462, 2543, 2588, 2623, 2627, 2652, 2652, 2654, 2666, 2690

I don't see a regression from db_basic_bench_pr to db_basic_bench_base ....

**Comment by ajkr** [comment]
> maybe I should try 1000 rep tomorrow (actually today's) morning.

Agreed, also remember to remove sources of known background noise

**Comment by hx235** [comment]
@ajkr finally the most welcomed regression 

Same set up as https://github.com/facebook/rocksdb/pull/11444#issuecomment-1573266783, except for the run command with --benchmark_repetitions=1000

 TEST_TMPDIR=/dev/shm ./db_basic_bench_{null_stat|base|pr} benchmark_filter=DBGet/comp_style:0/max_data:134217728/per_key_size:256/enable_statistics:1/negative_query:0/enable_filter:0/mmap:1/threads:1 --benchmark_repetitions=1000
 
 db_basic_bench_null_stat
 2175,2194,2200,2200,2219,2225,2229,2234,2243,2245
 
 db_basic_bench_base (regress 5-6% from db_basic_bench_null_stat)
2289,2314,2337,2340,2346,2347,2348,2361,2364,2369

db_basic_bench_pr (regress 4-5 % from db_basic_bench_base)
2372,2419,2421,2438,2439,2446,2458,2482,2487,2487

**Comment by hx235** [comment]
Ongoing: updating the PR summary for correct per result + guard new break down to kExceptDetailedTimers + add support for ReadAsync

File: utilities/write_batch_with_index/write_batch_with_index.cc [hunk 10]
```diff
@@ -531,7 +531,8 @@ Status WriteBatchWithIndex::GetFromBatchAndDB(
 
   // Did not find key in batch OR could not resolve Merges.  Try DB.
   if (!callback) {
-    s = db->Get(read_options, column_family, key, pinnable_val);
+    s = static_cast_with_check<DBImpl>(db->GetRootDB())
+            ->GetImpl(read_options, column_family, key, pinnable_val);
```
Comment (hx235) [inline utilities/write_batch_with_index/write_batch_with_index.cc:535]: I think it's safe to do so according the below lines

```
    DBImpl::GetImplOptions get_impl_options;
    get_impl_options.column_family = column_family;
    get_impl_options.value = pinnable_val;
    get_impl_options.callback = callback;
    s = static_cast_with_check<DBImpl>(db->GetRootDB())
            ->GetImpl(read_options, key, get_impl_options);
 ```

File: include/rocksdb/utilities/transaction.h [hunk 11]
```diff
@@ -333,8 +333,22 @@ class Transaction {
...
+        if (statuses[i].ok()) {
+          statuses[i] = s;
+        }
+      }
+      return;
+    }
+
     for (size_t i = 0; i < num_keys; ++i) {
-      statuses[i] = Get(options, column_family, keys[i], &values[i]);
+      statuses[i] = GetImpl(options, column_family, keys[i], &values[i]);
```
Comment (hx235) [inline include/rocksdb/utilities/transaction.h:351]: Refractory 1: 
Previous: 
```
Transaction::MultiGet() -> virtual Transaction::Get(PinnableSlice*) [overridden by child classes TransactionBase, WriteUn/PreparedTXN]-> Virtual Transaction::Get(string*) [only overridden by TransactionBase] -> TransactionBase::Get(string*) 
```
Now:
```
Transaction::MultiGet() -> virtual Transaction::GetImpl(PinnableSlice*) [overridden by child classes TransactionBase, WriteUn/PreparedTXN]-> Virtual Transaction::GetImpl(string*) [only overridden by TransactionBase] -> TransactionBase::GetImpl(string*) 
```

File: utilities/blob_db/blob_db_impl.cc [hunk 12]
```diff
@@ -1391,21 +1391,39 @@ std::vector<Status> BlobDBImpl::MultiGet(const ReadOptions& read_options,
...
+  }
+  bool snapshot_created = SetSnapshotIfNeeded(&complete_read_options);
+
   values->clear();
   values->reserve(keys.size());
   PinnableSlice value;
   for (size_t i = 0; i < keys.size(); i++) {
-    statuses.push_back(Get(ro, DefaultColumnFamily(), keys[i], &value));
+    statuses.push_back(
+        GetImpl(complete_read_options, DefaultColumnFamily(), keys[i], &value));
```
Comment (hx235) [inline utilities/blob_db/blob_db_impl.cc:1421]: Refactory 2: Call GetImpl(string*) instead of Get(string*) and ensure Get(string*) ist just GetImpl(string*) with io activity check

File: utilities/transactions/write_prepared_txn_db.cc [hunk 13]
```diff
@@ -319,11 +331,30 @@ std::vector<Status> WritePreparedTxnDB::MultiGet(
...
+  if (complete_read_options.io_activity == Env::IOActivity::kUnknown) {
+    complete_read_options.io_activity = Env::IOActivity::kMultiGet;
+  }
+
   values->resize(num_keys);
 
-  std::vector<Status> stat_list(num_keys);
   for (size_t i = 0; i < num_keys; ++i) {
-    stat_list[i] = this->Get(options, column_family[i], keys[i], &(*values)[i]);
+    stat_list[i] = this->GetImpl(complete_read_options, column_family[i],
```
Comment (hx235) [inline utilities/transactions/write_prepared_txn_db.cc:356]: Refactory 4:
Previous:
```
MultiGet()
-> virtual DB::Get(string*)
-> virtual DB::Get(PinnableSlice*) [overridden by WriteUnpreparedTxnDB]
-> WriteUnpreparedTxnDB::Get(PinnableSlice*) 
```

Now:
```
MultiGet()
-> WriteUnpreparedTxnDB::GetImpl(string*) [WriteUnpreparedTxnDB::GetImpl(string*) basically duplicates virtual DB::Get(string*) since I don't want to make a virtual DB::GetImpl() triggering more refactoring for name collision on XXDBImpl::GetImpl(). Some redundancy but keeping the scope of change small for now.]
-> WriteUnpreparedTxnDB::GetImpl(PinnableSlice*)
```

**Review by ajkr — COMMENTED** [review]
It looks like all the benchmarks in the description show regression, even though these stats should be disabled by default and some even explicitly specified "include_detailed_timers:0". Is this interpretation the same as yours, and if so, do you know why it regressed despite the `StatsLevel` guarding?

File: java/src/main/java/org/rocksdb/HistogramType.java [hunk 14]
```diff
@@ -175,6 +175,16 @@ public enum HistogramType {
...
 
+  FILE_READ_GET_MICROS((byte) 0x3D),
+
+  FILE_READ_MULTIGET_MICROS((byte) 0x3E),
+
+  FILE_READ_DB_ITERATOR_MICROS((byte) 0x3F),
+
+  FILE_READ_VERIFY_CHECKSUM_MICROS((byte) 0x41),
+
+  FILE_READ_VERIFY_FILE_CHECKSUMS_MICROS((byte) 0x42),
```
[the remark this answers is not in the document]
Comment (hx235) [reply java/src/main/java/org/rocksdb/HistogramType.java:186]: oops - was looking at NUM_SST_READ_PER_LEVEL((byte) 0x31) as example and thought 0xsomething0 was skipped intentionally until I now see `FLUSH_TIME((byte) 0x20),`

**Comment by hx235** [comment]
> It looks like all the benchmarks in the description show regression, even though these stats should be disabled by default and some even explicitly specified "include_detailed_timers:0". Is this interpretation the same as yours, and if so, do you know why it regressed despite the `StatsLevel` guarding?

Hi - I believe this is referring to AsyncRead() path. Yes, there is regression on this path because AsyncRead() didn't even collect sst.read.micros  and my latest iteration added that. I put the regression data there so that we can decide whether we should include adding it in this PR. I'm fine with both especially considering ReadAsync() is not yet very frequently used API so it's better to complete collecting metric before later.

My latest iteration can be summarized as https://github.com/facebook/rocksdb/pull/11444#issuecomment-1581798287 so it should not regression Get/MultiGet performance so I didn't re-do the perf test for them. I cleaned up the PR summary now but for some reason the previous Get performance data was not there anymore (probably accidentally deleted by me). Let me redo the perf test and provide the data in the summary then.

**Review by ajkr — APPROVED** [review]
OK I think the benchmarks look good assuming there's no regression shown for a `db_basic_bench_pr` - thanks! LGTM

File: db/db_iter.h [hunk 15]
```diff
@@ -392,6 +392,7 @@ class DBIter final : public Iterator {
   ColumnFamilyData* cfd_;
   const Slice* const timestamp_ub_;
   const Slice* const timestamp_lb_;
+  const Env::IOActivity io_activity_;
```
[the remark this answers is not in the document]
Comment (hx235) [reply db/db_iter.h:395]: Thanks - is there a document/rule-of-thumb you refer to when laying out Struct data members? I have not been exposed to that so would like to have one for my own reference.

[the remark this answers is not in the document]
Comment (ajkr) [reply db/db_iter.h:395]: The rule-of-thumb to minimize padding is to order the members by alignment, either ascending or descending is fine. In my memory descending is more common. This struct follows neither rule of thumb. Still, if you calculate how much padding is expected, appending a byte-aligned member after a bunch of `bool`s is less likely to add new padding than inserting it between two eight byte aligned fields.

[the remark this answers is not in the document]
Comment (hx235) [reply db/db_iter.h]: fixed

File: java/src/main/java/org/rocksdb/HistogramType.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (hx235) [reply java/src/main/java/org/rocksdb/HistogramType.java]: fixed

File: db/convenience.cc [hunk 16]
```diff
@@ -68,8 +68,8 @@ Status VerifySstFileChecksum(const Options& options,
       !kImmortal, false /* force_direct_prefetch */, -1 /* level */);
   reader_options.largest_seqno = largest_seqno;
   s = ioptions.table_factory->NewTableReader(
-      reader_options, std::move(file_reader), file_size, &table_reader,
```
[the remark this answers is not in the document]
Comment (hx235) [reply db/convenience.cc:71]: Thanks - fixed by renaming `FILE_READ_VERIFY_CHECKSUM_MICROS ` an related enum/string name to 
 `FILE_READ_DB_VERIFY_CHECKSUM_MICROS ` and adding sanitization to `VerifySstFileChecksum` so it does not update any stats by passing no IOActivity. It can be a follow-up to add a enum/stats about `VerifySstFileChecksum`
